DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp

```diff
@@ -0,0 +1,200 @@
+
+// Forward declare the kernel name in the global scope.
+// This is a FPGA best practice that reduces name mangling in the reports.
+class ConstructFromFloat;
+class ConstructFromACFixed;
+class CalculateWithFloat;
+class CalculateWithACFixed;
+
+// Not recommended Usage example:
+// Convert dynamic float value inside the kernel
+void TestConstructFromFloat(queue &q, const float &a,
+                            ac_fixed<20, 10, true, AC_RND, AC_SAT> &b) {
+  buffer<float, 1> inp_buffer(&a, 1);
+  buffer<ac_fixed<20, 10, true, AC_RND, AC_SAT>, 1> ret_buffer(&b, 1);
+
+  q.submit([&](handler &h) {
+    accessor in_acc{inp_buffer, h, read_only};
+    accessor out_acc{ret_buffer, h, write_only, no_init};
+
+    h.single_task<ConstructFromFloat>([=] {
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> t(in_acc[0]);
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> some_offset(0.5f);
+      out_acc[0] = t + some_offset;
+    });
+  });
+}
+
+// Recommended Usage example:
+// Convert dynamic float value outside the kernel
+void TestConstructFromACFixed(queue &q,
+                              const ac_fixed<20, 10, true, AC_RND, AC_SAT> &a,
+                              ac_fixed<20, 10, true, AC_RND, AC_SAT> &ret) {
+  buffer<ac_fixed<20, 10, true, AC_RND, AC_SAT>, 1> inp_buffer(&a, 1);
+  buffer<ac_fixed<20, 10, true, AC_RND, AC_SAT>, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor in_acc{inp_buffer, h, read_only};
+    accessor out_acc{ret_buffer, h, write_only};
+
+    h.single_task<ConstructFromACFixed>([=] {
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> t(in_acc[0]);
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> some_offset(0.5f);
+      out_acc[0] = t + some_offset;
+    });
+  });
+}
+
+void TestCalculateWithFloat(queue &q, const float &x, float &ret) {
+  buffer<float, 1> inp_buffer(&x, 1);
+  buffer<float, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp_buffer, h, read_only};
+    accessor res{ret_buffer, h, write_only, no_init};
+
+    h.single_task<CalculateWithFloat>([=] {
+      float sin_x = sinf(x[0]);
+      float cos_x = cosf(x[0]);
+      res[0] = sqrtf(sin_x * sin_x + cos_x * cos_x);
+    });
+  });
+}
+
+// clang-format off
+// Please refer to ac_fixed_math.hpp header file for fixed point math
+// functions' type deduction rule. In this case, following those rules:
+// I, W, S are input type template parameter (ac_fixed<W, I, S>)
+// rI, rW, rS are output type template parameter (ac_fixed<rW, rI, rS>)
+//* Function Name         Type Propagation Rule
+//* sqrt_fixed             rI = I, rW = W, rS = S
+//* sin_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+//* cos_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+// clang-format on
+void TestCalculateWithACFixed(queue &q, const fixed_10_3_t &x,
```

Comment:
Input called x vs a in above functions, please use the same convention for all these functions